// bkash-payment-api/dist/index.mjs
var __defProp = Object.defineProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// src/utils/diffSeconds.ts
function diffSeconds(date) {
  const diff = Date.now() - date;
  return diff / 1e3;
}

// src/utils/request.ts
import AbortController from "abort-controller";

// src/exceptions/bkashException.ts
var BkashException = class extends Error {
  constructor(message) {
    var _a;
    super(message);
    this.name = "BkashException";
    this.stack = (_a = this.stack) != null ? _a : new Error().stack;
  }
};

// src/utils/request.ts
function get(_0, _1) {
  return __async(this, arguments, function* (url, additionalHeaders, payload = {}) {
    const response = yield fetch(url, {
      method: "GET",
      headers: __spreadValues({
        "content-type": "application/json",
        Accept: "application/json"
      }, additionalHeaders),
      body: JSON.stringify(payload)
    });
    const parsed = yield response.json();
    if (parsed.errorMessage)
      throw new BkashException(parsed.errorMessage);
    return parsed;
  });
}
function post(_0) {
  return __async(this, arguments, function* (url, payload = {}, additionalHeaders) {
    const controller = new AbortController();
    const timeout = setTimeout(() => {
      controller.abort();
    }, 30 * 1e3);
    const response = yield fetch(url, {
      headers: __spreadValues({
        "content-type": "application/json",
        Accept: "application/json"
      }, additionalHeaders),
      body: JSON.stringify(payload),
      method: "POST"
      // signal: controller.signal,
    });
    clearTimeout(timeout);
    const parsed = yield response.json();
    if (parsed.errorMessage)
      throw new BkashException(parsed.errorMessage);
    return parsed;
  });
}

// src/bkashGateway.ts
var BkashGateway = class {
  /**
   *
   * @param config config object required by the `bkash-payment-gateway` package
   * @example
   * ```
   * const bkashConfig = {
   *   baseURL: 'https://tokenized.sandbox.bka.sh/v1.2.0-beta', //sandbox
   *   key: 'abcdxx2369',
   *   username: 'bkashTest',
   *   password: 'bkashPassword1',
   *   secret: 'bkashSup3rS3cRet',
   * }
   * const bkash = new BkashGateway(config)
   * ```
   *
   */
  constructor(config) {
    /**
     * Start the initial payment request
     *
     * @param paymentDetails Information required to start a payment flow
     *
     * @returns Promise of Bkash Create payment Response
     * @example
     * ```
     * const result = await bkash.createPayment({
     *   amount: 1000,
     *   orderID: 'ORD1020069',
     *   intent: 'sale',
     *   callBack: 'https://example.com/bkash/callback',
     * });
     * ```
     */
    this.createPayment = (paymentDetails) => __async(this, null, function* () {
      const { amount, intent, orderID, callBack } = paymentDetails;
      const payload = {
        orderId: orderID,
        amount: amount.toString(),
        intent,
        mode: "0011",
        currency: "BDT",
        merchantInvoiceNumber: orderID,
        payerReference: orderID,
        callbackURL: callBack
      };
      const headers = yield this.createTokenHeader();
      return yield post(
        `${this.baseURL}/tokenized/checkout/create`,
        payload,
        headers
      );
    });
    /**
     * Execute a payment after a user has completed bkash auth flow
     * @param paymentID - Payment ID to Execute
     * @example
     * ```
     * const result = await bkash.executePayment(paymentID);
     * ```
     */
    this.executePayment = (paymentID) => __async(this, null, function* () {
      try {
        const headers = yield this.createExecuteHeader();
        return yield post(
          `${this.baseURL}/tokenized/checkout/execute`,
          { paymentID },
          headers
        );
      } catch (error) {
        if (error instanceof BkashException) {
          throw error;
        }
        throw new BkashException(
          "Timeout of 30 Seconds Exceeded While Executing Payment, Please Query the Payment"
        );
      }
    });
    /**
     * Query Payment From Bkash
     * @param paymentID - Payment ID to Query
     *
     * @example
     * ```
     * const result = await bkash.queryPayment(paymentID);
     * ```
     */
    this.queryPayment = (paymentID) => __async(this, null, function* () {
      const headers = yield this.createTokenHeader();
      return yield get(
        `${this.baseURL}/tokenized/checkout/payment/status`,
        headers,
        {
          paymentID
        }
      );
    });
    /**
     * Search with a transaction ID
     * @param trxID - Transaction ID to Search
     *
     * @example
     * ```
     * const result = await bkash.searchTransaction('TRX22347463XX');
     * ```
     */
    this.searchTransaction = (trxID) => __async(this, null, function* () {
      return yield get(
        `${this.baseURL}/tokenized/checkout/general/searchTran`,
        yield this.createTokenHeader(),
        { trxID }
      );
    });
    /**
     * Refund a transaction
     * @param trxID - Transaction ID to Search
     * @deprecated
     * @example
     * ```
     * const refunTransactionData = {
     *  paymentID: '22423169',
     *  amount: '25.69', //do not add more than two decimal points
     *  trxID: 'TRX22347463XX',
     *  sku: 'SK256519',
     * };
     *
     * const result = await bkash.refundTransaction(refunTransactionData);
     * ```
     */
    this.refundTransaction = (refundInfo) => __async(this, null, function* () {
      return post(
        `${this.baseURL}/tokenized/checkout/payment/refund`,
        refundInfo,
        yield this.createTokenHeader()
      );
    });
    /**
     * Check Refund Status for a given paymentID and transaction ID
     * @param trxID transaction ID
     * @param paymentID payment ID
     * @deprecated
     * @example
     * ```
     * const result = await bkash.refundStatus('TRX22347463XX', '12437969');
     * ```
     */
    this.refundStatus = (trxID, paymentID) => __async(this, null, function* () {
      return yield post(
        `${this.baseURL}/tokenized/checkout/payment/refund`,
        { trxID, paymentID },
        yield this.createTokenHeader()
      );
    });
    this.createTokenHeader = (isBearer) => __async(this, null, function* () {
      const token = yield this.getToken();
      isBearer = isBearer != null ? isBearer : true;
      return {
        authorization: isBearer ? `Bearer ${token}` : token,
        "x-app-key": this.key,
        accept: "application/json",
        "content-type": "application/json"
      };
    });
    this.createExecuteHeader = () => __async(this, null, function* () {
      const token = yield this.getToken();
      return {
        Authorization: token,
        "x-app-key": this.key,
        accept: "application/json",
        "content-type": "application/json"
      };
    });
    this.getToken = () => __async(this, null, function* () {
      if (!this.token) {
        const { id_token: id_token2, refresh_token: refresh_token2, msg: msg2, status: status2 } = yield this.getInitialToken();
        if (status2 && msg2)
          throw new BkashException(msg2);
        this.token = id_token2;
        this.refreshToken = refresh_token2;
        this.tokenIssueTime = Date.now();
        return this.token;
      }
      const diff = diffSeconds(this.tokenIssueTime);
      if (diff < 3500) {
        return this.token;
      }
      const { id_token, refresh_token, msg, status } = yield this.newToken(
        this.refreshToken
      );
      if (status && msg)
        throw new BkashException(msg);
      this.token = id_token;
      this.refreshToken = refresh_token;
      this.tokenIssueTime = Date.now();
      return this.token;
    });
    this.getInitialToken = () => __async(this, null, function* () {
      const response = yield post(
        `${this.baseURL}/tokenized/checkout/token/grant`,
        {
          app_key: this.key,
          app_secret: this.secret
        },
        this.headers
      );
      if (response.status === "fail")
        throw new BkashException("Invalid API Credentials Provided");
      return response;
    });
    this.newToken = (refresh) => {
      return post(
        `${this.baseURL}/tokenized/checkout/token/refresh`,
        {
          app_key: this.key,
          app_secret: this.secret,
          refresh_token: refresh
        },
        this.headers
      );
    };
    this.validateConfig = (config) => {
      const { baseURL, key, password, secret, username } = config;
      if (!baseURL || baseURL === "")
        throw new BkashException("Invalid BaseURL provided");
      if (!key || key === "")
        throw new BkashException("Invalid API Key provided");
      if (!secret || secret === "")
        throw new BkashException("Invalid API secret provided");
      if (!username || username === "")
        throw new BkashException("Invalid API username provided");
      if (!password || password === "")
        throw new BkashException("Invalid API password provided");
    };
    if (Object.keys(config).length !== 5)
      throw new BkashException("Invalid Configuration provided");
    this.validateConfig(config);
    const { baseURL, key, password, secret, username } = config;
    this.baseURL = baseURL;
    this.key = key;
    this.secret = secret;
    this.headers = {
      username,
      password
    };
  }
};

// src/index.ts
var src_default = BkashGateway;
export {
  BkashException,
  BkashGateway,
  src_default as default
};
